lib/make-disk-image.nix

```diff
@@ -0,0 +1,56 @@
+{ nixosConfig
+, diskoLib
+, pkgs ? nixosConfig.pkgs
+, lib ? pkgs.lib
+, name ? "${nixosConfig.config.networking.hostName}-disko-images"
+}:
+let
+  cleanedConfig = diskoLib.testLib.prepareDiskoConfig nixosConfig.config diskoLib.testLib.devices;
+  systemToInstall = nixosConfig.extendModules {
+    modules = [{
+      disko.devices = lib.mkForce cleanedConfig.disko.devices;
+      boot.loader.grub.devices = lib.mkForce cleanedConfig.boot.loader.grub.devices;
+    }];
+  };
+in
+pkgs.vmTools.runInLinuxVM (pkgs.runCommand name {
+  buildInputs = with pkgs; [
+    systemdMinimal
+    nixos-install-tools
+    nix
+    utillinux
+  ];
+  preVM = ''
+    # TODO: get size either dynamically or from disko config
+    ${lib.concatMapStringsSep "\n" (disk: "truncate -s 15G ${disk.name}.raw") (lib.attrValues nixosConfig.config.disko.devices.disk)}
+  '';
+  postVM = ''
+    mkdir -p $out
+    ${lib.concatMapStringsSep "\n" (disk: "cp ${disk.name}.raw $out/${disk.name}.raw") (lib.attrValues nixosConfig.config.disko.devices.disk)}
+  '';
+  QEMU_OPTS = lib.concatMapStringsSep " " (disk: "-drive file=${disk.name}.raw,if=virtio,cache=unsafe,werror=report") (lib.attrValues nixosConfig.config.disko.devices.disk);
+
+  memSize = 1024;
+} ''
+  # running udev, stolen from stage-1.sh
+  echo "running udev..."
+  ln -sfn /proc/self/fd /dev/fd
+  ln -sfn /proc/self/fd/0 /dev/stdin
+  ln -sfn /proc/self/fd/1 /dev/stdout
+  ln -sfn /proc/self/fd/2 /dev/stderr
```

Comment:
Do we have other common devices such as /dev/null?

Reply:
dunno, this is taken from nixos stage-1